cpp/src/arrow/util/variant.h

```diff
@@ -17,17 +17,363 @@
+  return std::forward<Visitor>(visitor)(*static_cast<const T*>(ptr));
+}
+
+template <typename Visitor, typename T>
+decltype(std::declval<Visitor&&>()(std::declval<T*>())) visit_mutable_ptr(
+    Visitor&& visitor, void* ptr) {
+  return std::forward<Visitor>(visitor)(static_cast<T*>(ptr));
+}
+
+template <typename... T>
+struct variant_storage {
+  variant_storage() = default;
+  variant_storage(const variant_storage&) {}
+  variant_storage& operator=(const variant_storage&) { return *this; }
+  variant_storage(variant_storage&&) {}
+  variant_storage& operator=(variant_storage&&) { return *this; }
+  ~variant_storage() {
+    static_assert(offsetof(variant_storage, data_) == 0,
+                  "(void*)&variant_storage::data_ == (void*)this");
+  }
+
+  typename std::aligned_union<0, T...>::type data_;
+  uint8_t index_ = 0;
+};
+
+struct delete_copy_constructor {
+  template <typename>
+  struct type {
+    type() = default;
+    type(const type& other) = delete;
+    type& operator=(const type& other) = delete;
+  };
+};
+
+struct explicit_copy_constructor {
+  template <typename Copyable>
+  struct type {
+    type() = default;
+    type(const type& other) {
+      static_cast<Copyable*>(this)->copy(other, /*assignment=*/false);
+    }
+    type& operator=(const type& other) {
+      static_cast<Copyable*>(this)->copy(other, /*assignment=*/true);
+      return *this;
+    }
+  };
+};
+
+template <typename V, uint8_t I, typename...>
+struct member_constructor {
+  static void index_of() {}
+};
+
+template <typename V, uint8_t I, typename H, typename... T>
+struct member_constructor<V, I, H, T...> : member_constructor<V, I + 1, T...> {
+  static_assert(std::is_nothrow_move_constructible<H>::value &&
+                    std::is_nothrow_move_assignable<H>::value,
+                "All members must be nothrow move constructible and assignable.");
+
+  member_constructor() = default;
+
+  using member_constructor<V, I + 1, T...>::member_constructor;
+  using member_constructor<V, I + 1, T...>::operator=;
+  using member_constructor<V, I + 1, T...>::index_of;
+
+  explicit member_constructor(H value) {
+    new (this) H(std::move(value));
+    static_cast<V*>(this)->index_ = I;
+  }
+
+  member_constructor& operator=(H value) {
+    static_cast<V*>(this)->destroy();
+    new (this) H(std::move(value));
+    static_cast<V*>(this)->index_ = I;
+    return *this;
+  }
+
+  static constexpr std::integral_constant<uint8_t, I> index_of(const type_constant<H>&) {
+    return {};
+  }
+};
+
+}  // namespace detail
+
+template <typename... T>
+class Variant : detail::variant_storage<T...>,
+                detail::conditional_t<
+                    detail::all<(std::is_copy_constructible<T>::value &&
+                                 std::is_copy_assignable<T>::value)...>::value,
+                    detail::explicit_copy_constructor,
+                    detail::delete_copy_constructor>::template type<Variant<T...>>,
```

Comment:
We might want to borrow this idiom to disable copy construction for `Result<std::unique_ptr<T>>`